components/script/dom/servoparser/async_html.rs

```diff
@@ -0,0 +1,561 @@
+                    DOMString::from(data.data.clone().unwrap()),
+                    document);
+                self.nodes.borrow_mut().insert(node, JS::from_ref(pi.upcast()));
+            }
+        }
+    }
+}
+
+#[allow(unrooted_must_root)]
+impl TreeSink for Sink {
+    type Output = Self;
+    fn finish(self) -> Self { self }
+
+    type Handle = ParseNode;
+
+    fn get_document(&mut self) -> Self::Handle {
+        self.document_node.clone()
+    }
+
+    fn get_template_contents(&mut self, target: &Self::Handle) -> Self::Handle {
+        let data;
+        {
+            let parse_node_data = self.parse_node_data.borrow();
+            data = parse_node_data.get(&target.id).expect("Expected data of parse node").contents.clone();
+        }
+        if data.is_some() {
+            return data.clone().unwrap();
+        }
+
+        let node = self.new_parse_node();
+        let mut parse_node_data = self.parse_node_data.borrow_mut();
+        let mut data = parse_node_data.get_mut(&target.id).expect("Expected data of parse node");
+        data.contents = Some(node.clone());
+        self.enqueue(ParseOperation::GetTemplateContents(target.id, node.id));
+        node
+    }
+
+    fn same_node(&self, x: &Self::Handle, y: &Self::Handle) -> bool {
+        x.id == y.id
+    }
+
+    fn elem_name<'a>(&self, target: &'a Self::Handle) -> ExpandedName<'a> {
+        match target.qual_name {
+            Some(ref qual_name) => qual_name.expanded(),
+            None => panic!("Expected qual name of node!"),
+        }
+    }
+
+    fn same_tree(&self, x: &Self::Handle, y: &Self::Handle) -> bool {
+        let nodes = self.nodes.borrow();
+        let (id_x, id_y) = (x.id, y.id);
+
+        let x = nodes.get(&id_x).expect("Node not found");
+        // Update x's parent if no longer in tree
+        if !x.is_in_doc() {
+            self.parse_node_data.borrow_mut().get_mut(&id_x).expect("Node data not found").parent = None;
+        }
+
+        let y = nodes.get(&id_y).expect("Node not found");
+        // Update y's parent if no longer in tree
+        if !y.is_in_doc() {
+            self.parse_node_data.borrow_mut().get_mut(&id_y).expect("Node data not found").parent = None;
+        }
+
+        let x = x.downcast::<Element>().expect("Element node expected");
+        let y = y.downcast::<Element>().expect("Element node expected");
+        x.is_in_same_home_subtree(y)
+    }
+
+    fn create_element(&mut self, name: QualName, attrs: Vec<Attribute>, _flags: ElementFlags)
+        -> Self::Handle {
+        let mut node = self.new_parse_node();
+        node.qual_name = Some(name.clone());
+
+        let mut parse_node_data = self.parse_node_data.borrow_mut();
+        let mut node_data = parse_node_data.get_mut(&node.id).expect("Element does not exist!");
+        node_data.is_integration_point = attrs.iter()
+        .any(|attr| {
+            let attr_value = &String::from(attr.value.clone());
```

Comment:
I found that I could compute this value once and store it when I'm creating a new element (the value's needed in `is_mathml_annotation_xml_integration_point`). But now that I think about it, this could be a problem if the element's attributes related to this value are changed at any point of time *after* the element's created. Do you think I should change it to be synchronous ? (i.e, get the actual `Node` from the nodes map and compute the value from it).